pkg/inventory/model/client.go

```diff
@@ -0,0 +1,281 @@
+	for _, m := range r.models {
+		ddl, err := Table{}.DDL(m)
+		if err != nil {
+			panic(err)
+		}
+		statements = append(statements, ddl...)
+	}
+	for _, ddl := range statements {
+		_, err = db.Exec(ddl)
+		if err != nil {
+			Log.Trace(err)
+			db.Close()
+			return err
+		}
+	}
+
+	r.db = db
+
+	Log.Info("Database opened.", "path", r.path)
+
+	return nil
+}
+
+//
+// Close the database.
+// Optionally purge (delete) the DB.
+func (r *Client) Close(purge bool) error {
+	if r.db == nil {
+		return nil
+	}
+	err := r.db.Close()
+	if err != nil {
+		Log.Trace(err)
+		return err
+	}
+	r.db = nil
+	if purge {
+		os.Remove(r.path)
+	}
+
+	return nil
+}
+
+//
+// Get the model.
+func (r *Client) Get(model Model) error {
+	return Table{r.db}.Get(model)
+}
+
+//
+// List models.
+func (r *Client) List(model Model, options ListOptions, list interface{}) error {
+	mv := reflect.TypeOf(model)
+	switch mv.Kind() {
+	case reflect.Ptr:
+		mv = mv.Elem()
+	default:
+		return nil
+	}
+	lv := reflect.ValueOf(list)
+	lt := reflect.TypeOf(list)
+	switch lt.Kind() {
+	case reflect.Ptr:
+		lv = lv.Elem()
+		lt = lt.Elem()
+	default:
+		return nil
+	}
+	switch lv.Kind() {
+	case reflect.Slice:
+		l, err := Table{r.db}.List(model, options)
+		if err != nil {
+			Log.Trace(err)
+			return err
+		}
+		concrete := reflect.MakeSlice(lv.Type(), 0, 0)
+		for i := 0; i < len(l); i++ {
+			m := reflect.ValueOf(l[i]).Elem()
+			concrete = reflect.Append(concrete, m)
+		}
+		lv.Set(concrete)
+	}
+
+	return nil
+}
+
+//
+// Begin a transaction.
+// Example:
+//   tx, _ := client.Begin()
+//   defer tx.Rollback()
+//   client.Insert(model)
+//   client.Insert(model)
+//   tx.Commit()
+func (r *Client) Begin() (*Tx, error) {
+	r.Lock()
+	defer r.Unlock()
+	r.mutex.Lock()
+	tx, err := r.db.Begin()
+	if err != nil {
+		return nil, err
+	}
+	r.tx = tx
+	return &Tx{client: r, ref: tx}, nil
+}
+
+//
+// Insert the model.
+func (r *Client) Insert(model Model) error {
+	r.Lock()
+	defer r.Unlock()
+	model.SetPk()
+	table := Table{}
+	if r.tx == nil {
+		r.mutex.Lock()
+		defer r.mutex.Unlock()
+		table.Db = r.db
+	} else {
+		table.Db = r.tx
+	}
+
+	return table.Insert(model)
+}
+
+//
+// Update the model.
+func (r *Client) Update(model Model) error {
+	r.Lock()
+	defer r.Unlock()
+	model.SetPk()
+	table := Table{}
+	if r.tx == nil {
+		r.mutex.Lock()
+		defer r.mutex.Unlock()
+		table.Db = r.db
+	} else {
+		table.Db = r.tx
+	}
+
+	return table.Update(model)
+}
+
+//
+// Delete the model.
+func (r *Client) Delete(model Model) error {
+	r.Lock()
+	defer r.Unlock()
+	model.SetPk()
+	table := Table{}
+	if r.tx == nil {
+		r.mutex.Lock()
+		defer r.mutex.Unlock()
+		table.Db = r.db
+	} else {
+		table.Db = r.tx
+	}
+
+	return table.Delete(model)
+}
+
+//
+// Commit a transaction.
+// This MUST be preceeded by Begin() which returns
+// the `tx` transaction token.
+func (r *Client) commit(tx *Tx) error {
+	r.Lock()
+	defer r.Unlock()
+	if r.tx == nil || r.tx != tx.ref {
```

Comment:
I wonder if this branch ought to panic or return an error rather than silently returning.

Reply:
Yes, I think so.  Updated to return an error.